src/mono/msbuild/AndroidAppBuilder/Templates/runtime-android.c

```diff
@@ -0,0 +1,197 @@
+    return str ? mono_string_to_utf8 (str) : NULL;
+}
+
+void
+unhandled_exception_handler (MonoObject *exc, void *user_data)
+{
+    MonoClass *type = mono_object_get_class (exc);
+    char *type_name = strdup_printf ("%s.%s", mono_class_get_namespace (type), mono_class_get_name (type));
+    char *trace = fetch_exception_property_string (exc, "get_StackTrace", true);
+    char *message = fetch_exception_property_string (exc, "get_Message", true);
+    
+    LOG_ERROR("UnhandledException: %s %s %s", type_name, message, trace);
+
+    free (trace);
+    free (message);
+    free (type_name);
+    exit (1);
+}
+
+void
+log_callback (const char *log_domain, const char *log_level, const char *message, mono_bool fatal, void *user_data)
+{
+    LOG_INFO ("(%s %s) %s", log_domain, log_level, message);
+    if (fatal) {
+        LOG_ERROR ("Exit code: %d.", 1);
+        exit (1);
+    }
+}
+
+int
+mono_mobile_runtime_init (void)
+{
+    // for now, only Invariant Mode is supported (FIXME: integrate ICU)
+    setenv ("DOTNET_SYSTEM_GLOBALIZATION_INVARIANT", "1", true);
+    // uncomment for debug output:
+    //
+    // setenv ("MONO_LOG_LEVEL", "debug", TRUE);
+    // setenv ("MONO_LOG_MASK", "all", TRUE);
+
+    bool wait_for_debugger = false;
+    chdir (bundle_path);
+
+    // TODO: set TRUSTED_PLATFORM_ASSEMBLIES, APP_PATHS and NATIVE_DLL_SEARCH_DIRECTORIES
+    monovm_initialize(0, NULL, NULL);
+
+    mono_debug_init (MONO_DEBUG_FORMAT_MONO);
+    mono_install_assembly_preload_hook (assembly_preload_hook, NULL);
+    mono_install_unhandled_exception_hook (unhandled_exception_handler, NULL);
+    mono_trace_set_log_handler (log_callback, NULL);
+    mono_set_signal_chaining (true);
+    mono_set_crash_chaining (true);
+
+    if (wait_for_debugger) {
+        char* options[] = { "--debugger-agent=transport=dt_socket,server=y,address=0.0.0.0:55555" };
+        mono_jit_parse_options (1, options);
+    }
+    mono_jit_init_version ("mono.app", "mobile");
+
+    const char* executable = "%EntryPointLibName%";
+    MonoAssembly *assembly = load_assembly (executable, NULL);
+    assert (assembly);
+    LOG_INFO ("Executable: %s", executable);
+
+    char *managed_argv [1];
+    managed_argv[0] = bundle_path;
+
+    int res = mono_jit_exec (mono_domain_get (), assembly, 1, managed_argv);
+    LOG_INFO ("Exit code: %d.", res);
+    return res;
+}
+
+static void
+strncpy_str (JNIEnv *env, char *buff, jstring str, int nbuff)
+{
+    jboolean isCopy = 0;
+    const char *copy_buff = (*env)->GetStringUTFChars (env, str, &isCopy);
+    strncpy (buff, copy_buff, nbuff);
+    if (isCopy)
+        (*env)->ReleaseStringUTFChars (env, str, copy_buff);
+}
+
+int
+Java_net_dot_MonoRunner_initRuntime (JNIEnv* env, jobject thiz, jstring j_files_dir)
+{
+    char file_dir[2048];
+    strncpy_str (env, file_dir, j_files_dir, sizeof(file_dir));
+    bundle_path = file_dir;
+    setenv ("HOME", bundle_path, true);
```

Comment:
is this what XA does too?

Reply:
@akoeplinger yes as far as I understand

Reply:
ok yeah I think I found it: https://github.com/xamarin/xamarin-android/blob/fb727f0a64ac2932b96a117c4a26786eac8e0f6c/src/monodroid/jni/monodroid-glue.cc#L1450-L1451

We might want to set TMPDIR too but that can be in another PR ;)

Reply:
Turned out `TMPDIR` is also needed (crashed without it)